### PR TITLE
Fix #17248 - Support UUID data type for MariaDB

### LINKED ISSUE #17248: please support uuid data type
### Is your feature request related to a problem? Please describe.

[A clear and concise description of what the problem is. Ex. I'm always frustrated when [...]](https://giters.com/michaeldyrynda/laravel-efficient-uuid/issues/42)

### Describe the solution you'd like

The following picture shows that the datatype is uuid, which is correct
image

But the datatype shown in the figure below is INT.
image

### Describe alternatives you've considered

Please support UUID type .

### Additional context

- phpMyAdmin version: 5.1.1

### PR DESCRIPTION
**Pull request by sureemo**
### Description

Support UUID data type for `MariaDB` since version `10.7.0`

Reference issue: https://github.com/phpmyadmin/phpmyadmin/issues/17248

Fixes: #17248

**Screenshots**

Screen Shot 2565-10-10 at 22 01 44

Screen Shot 2565-10-10 at 22 02 20

Signed-off-by: Mo Sureerat <sureemo@gmail.com>

### CODE REVIEW

File: libraries/classes/Types.php [hunk 1]
```diff
@@ -906,4 +932,19 @@ public function getIntegerRange($type, $signed = true)
...
         ];
     }
+
+    /**
+     * Check whether the database support uuid
+     * true if uuid is supported
+     *
+     * @return bool enabled status
+     */
+    private function isUUIDEnabled(): bool
```
Comment (williamdes) [inline libraries/classes/Types.php:942]: There is a class where such logic resides, in the folder query.
The class Compatibility

Comment (sureemo) [reply libraries/classes/Types.php]: thank you for suggestion, the method is now moved.

File: libraries/classes/DatabaseInterface.php [hunk 2]
```diff
@@ -2282,6 +2282,17 @@ public function isPercona(): bool
         return $this->isPercona;
     }
 
+    /**
+     * Set server version (initially to use in unit test)
+     */
+    public function setDatabaseVersion(string $database, string $version): void
```
Comment (MauricioFauth) [inline libraries/classes/DatabaseInterface.php:2288]: I think it's better to use/improve the `DatabaseInterface::setVersion` function instead.

**Review by MauricioFauth — CHANGES_REQUESTED** [review]

File: libraries/classes/DatabaseInterface.php (hunk 2 above)
Comment (MauricioFauth) [reply libraries/classes/DatabaseInterface.php:2288]: You can backport this function to QA_5_2:

https://github.com/phpmyadmin/phpmyadmin/blob/6785c97d22e764bb87cfa86eae199d5f7909297a/libraries/classes/DatabaseInterface.php#L2037-L2044

Comment (sureemo) [reply libraries/classes/DatabaseInterface.php]: thank you for suggestions, i have moved to `setVersion` but use logic from `postConnect`. please feel free to comment.

File: libraries/classes/DatabaseInterface.php [hunk 3]
```diff
@@ -2282,6 +2273,29 @@ public function isPercona(): bool
...
+     * @phpstan-param array<array-key, mixed> $version
+     */
+    public function setVersion(array $version): void
+    {
+        $this->versionString = $version['@@version'] ?? '';
+        $this->versionInt = Utilities::versionToInt($this->versionString);
+        $this->versionComment = $version['@@version_comment'] ?? '';
+
+        if (stripos($this->versionString, 'mariadb') !== false) {
+            $this->isMariaDb = true;
```
Comment (williamdes) [inline libraries/classes/DatabaseInterface.php:2289]:
```suggestion
            $this->isMariaDb = true;
            return;
```
If isPercona is set to false by default we can quick exit here

File: libraries/classes/Query/Compatibility.php [hunk 4]
```diff
@@ -200,6 +202,19 @@ public static function isVirtualColumnsSupported(int $serverVersion): bool
...
+     * Check whether the database support uuid
+     * true if uuid is supported
+     */
+    public static function isUUIDSupported(DatabaseInterface $dbi): bool
+    {
+        if (! $dbi->isMariaDB()) {
+            return false;
+        }
+
+        return $dbi->getVersion() >= self::UUID_SUPPORT_MARIADB_VERSION;
```
Comment (williamdes) [inline libraries/classes/Query/Compatibility.php:215]: I think we can do like other methods and not use a const

File: test/classes/TypesTest.php [hunk 5]
```diff
@@ -335,40 +335,39 @@ public function providerFortTestGetFunctionsClass(): array
...
-                    'SOUNDEX',
-                    'SPACE',
-                    'TRIM',
-                    'UNCOMPRESS',
-                    'UNHEX',
-                    'UPPER',
-                    'USER',
-                    'UUID',
-                    'VERSION',
+                    '0' => 'AES_DECRYPT',
```
Comment (williamdes) [inline test/classes/TypesTest.php:338]: I think you could avoid this by doing some `array_values` when you remove the uuid key

**Review by williamdes — APPROVED** [review]

File: libraries/classes/DatabaseInterface.php (hunk 3 above)
Comment (sureemo) [reply libraries/classes/DatabaseInterface.php]: Thank you to comment on this code, I think should it refresh value both `isMariaDb` and `isPercona` .. so remove condition check is better.

File: libraries/classes/Query/Compatibility.php (hunk 4 above)
Comment (sureemo) [reply libraries/classes/Query/Compatibility.php]: updated as suggestion. thank you.

File: test/classes/TypesTest.php (hunk 5 above)
Comment (sureemo) [reply test/classes/TypesTest.php]: improve the code as suggestion. thank you.

**Review by williamdes — APPROVED** [review]

File: libraries/classes/Query/Compatibility.php [hunk 6]
```diff
@@ -200,6 +200,19 @@ public static function isVirtualColumnsSupported(int $serverVersion): bool
         return false;
     }
 
+    /**
+     * Check whether the database support uuid
+     * true if uuid is supported
```
Comment (kamil-tekiela) [inline libraries/classes/Query/Compatibility.php:205]:
```suggestion
     * Check whether the database supports UUID data type
```

File: libraries/classes/Query/Compatibility.php [hunk 7]
```diff
@@ -200,6 +200,19 @@ public static function isVirtualColumnsSupported(int $serverVersion): bool
...
+     */
+    public static function isUUIDSupported(DatabaseInterface $dbi): bool
+    {
+        if (! $dbi->isMariaDB()) {
+            return false;
+        }
+
+        return $dbi->getVersion() >= 100700; // 10.7.0
```
Comment (kamil-tekiela) [inline libraries/classes/Query/Compatibility.php:213]:
```suggestion
        return $dbi->isMariaDB() && $dbi->getVersion() >= 100700; // 10.7.0
```

**Review by kamil-tekiela — CHANGES_REQUESTED** [review]

**Review by MauricioFauth — APPROVED** [review]

**Review by williamdes — APPROVED** [review]

**Comment by williamdes** [comment]
Can I merge your work @sureemo ?

**Comment by sureemo** [comment]
> Merge branch 'QA_5_2' into feature/issue-17248-add-uuid

of course. but there are some tests fail .. about context keywords .. still not find the root cause.

**Comment by sureemo** [comment]
@williamdes build checks are green now. ready to be reviewed and merged. thank you ^_^

**Review by MauricioFauth — APPROVED** [review]

**Review by williamdes — APPROVED** [review]

**Comment by williamdes** [comment]
Fixes: #17248

**Pull request merged**